Kubernetes v1.32 sneak peek blog (#48585)
kubernetes/website: content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md

---
layout: blog
title: 'Kubernetes v1.32 sneak peek'
date: 2024-11-04
slug: kubernetes-1-32-upcoming-changes
author: >
  Matteo Bianchi,
  Edith Puclla,
  William Rizzo,
  Ryota Sawada,
  Rashan Smith
---

> As we get closer to the release date for Kubernetes v1.32, the project develops and matures, features may be deprecated, removed, or replaced with better ones for the project's overall health. 

Author: Implemented a slightly different change, thank you!

> This blog outlines some of the planned changes for the Kubernetes v1.32 release, that the release team feels you should be aware of, for the continued mantainance of your Kubernetes enviroment and keeping up to date with the latest changes. Information listed below is based on the current status of the v1.32 release and may change before the actual release date. 

Author: Solved, thanks!

> This blog outlines some of the planned changes for the Kubernetes v1.32 release, that the release team feels you should be aware of, for the continued mantainance of your Kubernetes enviroment and keeping up to date with the latest changes. Information listed below is based on the current status of the v1.32 release and may change before the actual release date. 

Author: Outdated, already implemented the fixes, but still thank you for pointing these out!

> ### The Kubernetes API Removal and Deprecation process

Reviewer: Please make this sentence case.

Author: Solved!

> The Kubernetes project has a well-documented [deprecation policy](/docs/reference/using-api/deprecation-policy/) for features. This policy states that stable APIs may only be deprecated when a newer, stable version of that same API is available and that APIs have a minimum lifetime for each stability level. A deprecated API has been marked for removal in a future Kubernetes release, it will continue to function until removal (at least one year from the deprecation), but usage will result in a warning being displayed. Removed APIs are no longer available in the current version, at which point you must migrate to using the replacement.

Author: Implemented a slightly different change, thank you!

> The Kubernetes project has a well-documented [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) for features. This policy states that stable APIs may only be deprecated when a newer, stable version of that same API is available and that APIs have a minimum lifetime for each stability level. A deprecated API has been marked for removal in a future Kubernetes release, it will continue to function until removal (at least one year from the deprecation), but usage will result in a warning being displayed. Removed APIs are no longer available in the current version, at which point you must migrate to using the replacement.

Reviewer 2: We use site-relative link where possible.

```suggestion
The Kubernetes project has a well-documented [deprecation policy](/docs/reference/using-api/deprecation-policy/) for features. This policy states that stable APIs may only be deprecated when a newer, stable version of that same API is available and that APIs have a minimum lifetime for each stability level. A deprecated API has been marked for removal in a future Kubernetes release, it will continue to function until removal (at least one year from the deprecation), but usage will result in a warning being displayed. Removed APIs are no longer available in the current version, at which point you must migrate to using the replacement.
```

* Generally available (GA) or stable API versions may be marked as deprecated but must not be removed within a major version of Kubernetes.

* Beta or pre-release API versions must be supported for 3 releases after the deprecation.

> * Alpha or experimental API versions may be removed in any release without prior deprecation notice, this process can become a withdrawal in cases where a different implementation for the same feature is already in place.

Author: Done, thanks!

> Whether an API is removed as a result of a feature graduating from beta to stable or because that API simply did not succeed, all removals comply with this deprecation policy. Whenever an API is removed, migration options are communicated in the [documentation](/docs/reference/using-api/deprecation-guide/).

Author: Implemented, thanks!

> Whether an API is removed as a result of a feature graduating from beta to stable or because that API simply did not succeed, all removals comply with this deprecation policy. Whenever an API is removed, migration options are communicated in the [documentation](https://kubernetes.io/docs/reference/using-api/deprecation-guide/).

Reviewer 2: ```suggestion
Whether an API is removed as a result of a feature graduating from beta to stable or because that API simply did not succeed, all removals comply with this deprecation policy. Whenever an API is removed, migration options are communicated in the [documentation](/docs/reference/using-api/deprecation-guide/).
```

## Note on the withdrawal of the old DRA implementation

> This enhancement [#3063](https://github.com/kubernetes/enhancements/issues/3063) initially introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.

Author: Implemented, thanks!

> However, in Kubernetes v1.32, such enhancement is withdrawn. Related code will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 

Reviewer 3: ```suggestion
However, in Kubernetes v1.32, this approach to DRA has been significantly changed. Code related to the original implementation will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 
```

Just a suggestion on wording, but I do think this paragraph needs to be clarified.

When reading through this, I read:

> such enhancement is withdrawn.

"Oh, DRA has been removed? I didn't realize that." Then:

> Even more DRA enhancements

"Wait, I thought it just said DRA was removed?"

Author: Good point, implemented your suggestion!

Thank you 😄

> However, in Kubernetes v1.32, it is withdrawn and related code will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 

Author: Implemented a slightly different change, thank you!

> The decision to remove this feature originated from its incompatibility with cluster autoscaling, as resource availability was made non-transparent, complicating decision making for both Cluster Autoscaler and controllers. 

Author: Implemented, thanks!

The newly added Structured Parameter model substitutes the functionality.

This removal allows Kubernetes to handle new hardware requirements and resource claims more predictably, bypassing the complexities of back and forth API calls to the kube-apiserver.

Please also see the enhancement issue [#3063](https://github.com/kubernetes/enhancements/issues/3063) to find out more.

## Sneak peek of Kubernetes v1.32

### Even more DRA enhancements!

> In this release, like the previous one, the Kubernetes project continues to introduce enhancements to the Dynamic Resource Allocation (DRA), which is a key component of the Kubernetes resource management system. These enhancements aim to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such as GPUs, FPGAs and network adapters. The enhancements for this release include:

Author: Implemented, thank you!

> - Partionable devices support: KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815)

Reviewer: Spelling nit
```suggestion
- Partitionable devices support: KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815)
```

Author: Thanks, fixed it!

- Resource Claim Status and standardized network interface data: [KEP #4817](https://github.com/kubernetes/enhancements/issues/4817)

> - Structured parameters: KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381)

Reviewer 4: ```suggestion
- Structured parameters: [KEP #4381](https://github.com/kubernetes/enhancements/issues/4381)
```
To maintain consistency in KEP number formatting with other points

- Prioritized Alternatives in Device Requests: [KEP #4816](https://github.com/kubernetes/enhancements/issues/4816)
- Resource Health Status in Pod Status: [KEP #4680](https://github.com/kubernetes/enhancements/issues/4680)

**Partitionable devices support**

KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815) introduces improvements to Kubernetes' Dynamic Resource Allocation (DRA) framework, enabling more efficient support for high-performance partitionable devices like GPUs, FPGAs, and network adapters. Currently, these devices are treated as whole units, potentially leading to resource inefficiencies when workloads require only a fraction of a device’s capacity. 

> By extending the DRA framework, this enhancement allows workloads to request and consume portions of these resources, increasing utilization and scheduling flexibility. The proposed enhancement also provides primitives for representing both full devices and their partitions in a more compact manner, supporting the recent transition from what was called "classic DRA" to the "structured parameters" approach. 

Author: Implemented!

> The newly proposed extensions, fully transparent to the end-user, enable vendors to provision “overlapping” partitions, ensuring that the scheduler does not allocate conflicting partitions. This approach allows vendors to dynamically create new partitions after allocation, eliminating the need for pre-partitioned configurations, while still supporting device partition selection through existing mechanisms like `ResourceClaim` and `ResourceSlice` in Kubernetes v1.31.

Author: Implemented!

**Resource Claim status and standardized network interface data**

KEP [#4817](https://github.com/kubernetes/enhancements/issues/4817) adds driver-owned fields in `ResourceClaim.Status` with the possibility of having standardized network interface data.
This proposal enhances the `ResourceClaim.Status` field by adding a new sub-field: `Devices`.
The new entry allows to report driver-specific device status data for each allocated device in a `ResourceClaim`. Allowing the drivers to report such data will improve observability and troubleshooting, as well as enabling new functionalities such as, for example, network services (in case the IP addresses of a network device are reported).

**DRA Structured Parameters** 

> The original dynamic resource allocation (DRA) uses claim and class parameters that are opaque to Kubernetes. KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) introduces structured parameters so that kube-scheduler and Cluster Autoscaler can handle a claim allocation themselves without relying on a third-party driver.

Author: Implemented, thank you!

> As users are increasingly adopting Kubernetes as their management solution for batch-processing workloads and edge computing, the need of exposing specialized hardware to Pods is increasing.

Author: Implemented, thanks!

> Such workloads no longer need just RAM and CPU - so this KEP introduced a new API for describing which of these specialized resources a Pod needs, for example devices like a GPU or other kinds of accelerators.

Author: Implemented, thank you!

**Prioritized alternatives in device requests** 

KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) adds support for a prioritized list of selection criteria, attached to a device request in a `ResourceClaim`.
The DRA's Structured Parameters feature has added the ability to make requests for very specific types of devices using a `ResourceClaim`.

> However, the current API did not allow the user to indicate any priority when multiple types or configurations of devices might have met the needs of the workload. This feature allows the user to specify alternative requests that satisfy the workloads need, giving more flexibility to the scheduler in the workloads' scheduling phase.

Author: Implemented, thanks!

**Add resource health status to the Pod status**

> When a Pod is using a device that has failed or is temporarily unhealthy, it is difficult to know. KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680) makes troubleshooting of Pod crashes easier by exposing device health via `Pod Status`.

Reviewer 5: ```suggestion
It isn't easy to know when a Pod uses a device that has failed or is temporarily unhealthy. KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680) makes troubleshooting Pod crashes easier by exposing device health via `Pod Status`.
```

Author: Implemented, thank you!

### Windows strikes back!

> KEP [#4802](https://github.com/kubernetes/enhancements/issues/4802) adds support for graceful shutdown of Windows nodes in Kubernetes clusters.

Reviewer 5: ```suggestion
KEP [#4802](https://github.com/kubernetes/enhancements/issues/4802) adds support for the graceful shutdown of Windows nodes in Kubernetes clusters.
```

Author: Implemented a slightly different change, thanks!

> Currently, Kubernetes provides graceful node shutdown functionality for Linux nodes but lacked equivalent support for Windows. This enhancement aims to enable the kubelet on Windows nodes to handle system shutdown events properly. By doing so, it ensures that Pods running on Windows nodes are gracefully terminated, allowing workloads to be rescheduled without disruption. This improvement enhances the reliability and stability of clusters that include Windows nodes, especially during a planned maintenance or any system updates.

Reviewer 5: ```suggestion
Currently, Kubernetes provides graceful node shutdown functionality for Linux nodes but lacks equivalent support for Windows. This enhancement aims to enable the kubelet on Windows nodes to handle system shutdown events properly. Doing so ensures that Pods running on Windows nodes are gracefully terminated, allowing workloads to be rescheduled without disruption. This improvement enhances the reliability and stability of clusters that include Windows nodes, especially during planned maintenance or system updates.
```

KEP [#4885](https://github.com/kubernetes/enhancements/issues/4885) adds CPU and memory affinity support to Windows nodes in Kubernetes, thus improving performance for workloads that benefit from cache vicinity and NUMA optimizations, this effort also aligns Windows capabilities with Linux nodes.

> 

Reviewer 6: A few more exciting things from SIG Node around quility of life and small improvements:

1. https://github.com/kubernetes/kubernetes/pull/127566 - support for systemd watchdog is implemented for the kubelet. 
2. Better message in image pull backoff message: https://github.com/kubernetes/kubernetes/pull/127918

Also we are preparing to the major GA in 1.33 - tracking last remaining items and feedback for sidecar containers KEP: https://github.com/kubernetes/enhancements/issues/753#issuecomment-2350136594

Author: Hey @SergeyKanzhelev 👋 

Thanks for pointing these out, amazing enhancements _huge fan of the better image pull backoff error btw_
Just so you know this blog being the sneak peek means that the KEP list chosen to be highlighted is non-exhaustive by definition, it's an arbitrary subset by the Release Team (Enhancements, Comms).

You might want to add these to the [Release Highlights discussion](https://github.com/kubernetes/sig-release/discussions/2639) for SIG Node instead, those and any KEPs highlighted by the SIG leadership will be included in the final Release Blog!

Reviewer 6: sounds good. https://github.com/kubernetes/sig-release/discussions/2639#discussioncomment-11093925

Author: Amazing, thank you!

> KEP [#4885](https://github.com/kubernetes/enhancements/issues/4885) adds CPU and memory affinity support to Windows nodes in Kubernetes, thus improving performance for workloads that benefit from cache vicinity and NUMA optimizations, this effort also aligns Windows capabilities with Linux nodes.

Reviewer 5: ```suggestion
KEP [#4885](https://github.com/kubernetes/enhancements/issues/4885) adds CPU and memory affinity support to Windows nodes in Kubernetes, thus improving performance for workloads that benefit from cache vicinity and NUMA optimizations; this effort also aligns Windows capabilities with Linux nodes.
```

Author: Implemented, thanks!

> ### Almost every ASCII character in env variables - KEP [#4369](https://github.com/kubernetes/enhancements/issues/4369)

Reviewer: I'd rephrase it to the below since the existing sentence sounds incomplete. WDYT?

```suggestion
### Allow special characters in environment variables - KEP [#4369](https://github.com/kubernetes/enhancements/issues/4369)
```

> With the graduation of KEP-4369 to beta, Kubernetes now allows almost all printable ASCII characters (excluding "=") to be used as environment variable names. This change addresses the limitations previously imposed on variable naming, facilitating broader adoption of Kubernetes by accommodating various application needs. The relaxed validation will be enabled by default via `RelaxedEnvironmentVariableValidation` feature gate, ensuring that users can easily utilize environment variables without strict constraints, enhancing flexibility for developers working with applications like .NET Core that require special characters in their configurations.

Reviewer 5: ```suggestion
With the graduation of KEP-4369 to beta, Kubernetes now allows almost all printable ASCII characters (excluding "=") to be used as environment variable names. This change addresses the limitations previously imposed on variable naming, facilitating the broader adoption of Kubernetes by accommodating various application needs. The relaxed validation will be enabled by default via the `RelaxedEnvironmentVariableValidation` feature gate, ensuring that users can easily utilize environment variables without strict constraints, enhancing flexibility for developers working with applications like .NET Core that require special characters in their configurations.
```

Author: Implemented, thank you!

### Field status.hostIPs added for Pod - KEP [#2681](https://github.com/kubernetes/enhancements/issues/2681)

> This enhancement introduces a new field, status.hostIPs, to the Kubernetes Pod API, enabling support for multiple IP addresses (IPv4 and IPv6) assigned to a node. Previously, the Pod status only included a singular HostIP field, which limited address configurations, particularly in dual-stack networks. This enhancement allows the API to store multiple IP addresses in an array, accessible via the Downward API, which applications can leverage for environments with complex IP requirements, such as IPv6 migrations. 

Reviewer: ```suggestion
This enhancement introduces a new field, `status.hostIPs`, to the Kubernetes Pod API, enabling support for multiple IP addresses (IPv4 and IPv6) assigned to a node. Previously, the Pod status only included a singular HostIP field, which limited address configurations, particularly in dual-stack networks. This enhancement allows the API to store multiple IP addresses in an array, accessible via the Downward API, which applications can leverage for environments with complex IP requirements, such as IPv6 migrations. 
```

Author: Another fix in, thank you!

The feature is controlled by the PodHostIPs feature gate, supporting flexibility for gradual rollout.

### Make Kubernetes aware of the LoadBalancer behavior - [KEP #1860](https://github.com/kubernetes/enhancements/issues/1860)

> KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. The existing behavior of kube-proxy is preserved when using `"VIP"`, where kube-proxy handles the load balancing. The use of `"Proxy"` results in traffic sent directly to the load balancer, providing cloud providers greater control over relying on kube-proxy.

Reviewer 7: So the KEP is maybe not super clear about this, but this isn't an end-user feature; the feature is there so that _people implementing cloud providers_ can better describe how their load balancers work, so that kube-proxy can integrate with them better, and then the end-user experience is just "some kinds of load balancers work better on some clouds than they used to" but the end user doesn't actually have to do anything to get that improvement. (So from the end user perspective, this isn't a new feature, it's just a bugfix.)

(And I'm not sure what cloud providers needed this, and which ones have adopted it...)

Author: Clarified adding some context for the end user, thanks for your contribution!

> KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. The existing behavior of kube-proxy is preserved when using `"VIP"`, where kube-proxy handles the load balancing. The use of `"Proxy"` results in traffic sent directly to the load balancer, providing cloud providers greater control over relying on kube-proxy.

Reviewer 5: ```suggestion
KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. The existing behavior of kube-proxy is preserved when using `"VIP"`, where kube-proxy handles the load balancing. Using `"Proxy"` results in traffic sent directly to the load balancer, providing cloud providers greater control over relying on kube-proxy.
```

Author: Implemented, thank you!

### Support PSI based on cgroupv2 - [KEP #4205](https://github.com/kubernetes/enhancements/issues/4205)

> This KEP adds support in the Kubelet to read Pressure Stall Information (PSI) metrics for CPU, Memory and I/O resources exposed from cAdvisor and runc. Pressure metrics are like a barometer that can warn about impending resource shortages on a node. This will enable the Kubelet to report node conditions, which can be utilized to prevent scheduling of Pods on nodes experiencing significant resource constraints.

Reviewer 5: ```suggestion
This KEP adds support in the Kubelet to read Pressure Stall Information (PSI) metrics for CPU, Memory, and I/O resources exposed from cAdvisor and runc. Pressure metrics are like a barometer that can warn about impending resource shortages on a node. This will enable the Kubelet to report node conditions, which can be utilized to prevent the scheduling of Pods on nodes experiencing significant resource constraints.
```

Author: Implemented, thank you!

### Retry generate name for resources - [KEP #4420](https://github.com/kubernetes/enhancements/issues/4420)

> This enhancement improves the handling of name conflicts for Kubernetes resources created with the `generateName` field. Previously, if a name conflict occurred, the API server returned a 409 error, and clients had to manually retry the request. With this update, the API server now automatically retries generating a new name up to seven times in case of a conflict. This significantly reduces the chances of collision, ensuring smooth generation of up to 1 million names with less than a 0.1% probability of a conflict, providing more resilience for large-scale workloads.

Reviewer 5: ```suggestion
This enhancement improves handling name conflicts for Kubernetes resources created with the `generateName` field. Previously, if a name conflict occurred, the API server returned a 409 error, and clients had to manually retry the request. With this update, the API server automatically retries, generating a new name up to seven times in case of a conflict. This significantly reduces the chances of collision, ensuring smooth generation of up to 1 million names with less than a 0.1% probability of a conflict, providing more resilience for large-scale workloads.
```

Author: Solved with a slightly different change, thank you!

### Mutating admission policies - [KEP #3962](https://github.com/kubernetes/enhancements/issues/3962)

> This enhancement introduces "Mutating Admission Policies" using CEL (Common Expression Language), allowing Kubernetes users to define in-line mutating policies without the need for complex admission webhooks. 

Reviewer 5: ```suggestion
This enhancement introduces "Mutating Admission Policies" using CEL (Common Expression Language), allowing Kubernetes users to define in-line mutating policies without complex admission webhooks. 
```

Author: Implemented, thanks!

> This enhancement simplifies common mutating tasks—such as adding labels, setting fields, or injecting sidecars—by enabling policies to be directly defined within the Kubernetes API server. With CEL's object instantiation and Server-Side Apply merge algorithms, these policies can minimize reinvocation, reduce latency, and eliminate the overhead associated with webhook management. This new approach enhances Kubernetes extensibility by reducing operational complexity, also provides a scalable, built-in solution for secure and efficient policy management across diverse use cases like GitOps and CI/CD, making Kubernetes admission control more accessible and powerful for developers and administrators alike.

Reviewer 5: ```suggestion
This enhancement simplifies common mutating tasks—such as adding labels, setting fields, or injecting sidecars—by enabling policies to be directly defined within the Kubernetes API server. With CEL's object instantiation and Server-Side Apply merge algorithms, these policies can minimize reinvocation, reduce latency, and eliminate the overhead associated with webhook management. This new approach enhances Kubernetes extensibility by reducing operational complexity. Also, it provides a scalable, built-in solution for secure and efficient policy management across diverse use cases like GitOps and CI/CD, making Kubernetes admission control more accessible and powerful for developers and administrators alike.
```

Author: Solved with a slightly different change!

### Control over terminating Pods in Deployments - KEP [#3973](https://github.com/kubernetes/enhancements/issues/3973)

> This enhancement adds a new field `.spec.PodReplacementPolicy` to Deployments, to specify whether new Pods can be scheduled only after old Pods have finished terminating, or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can potentially help avoid problems with excessive resource usage or slow scheduling during rollouts and scaling.

Reviewer 8: ```suggestion
This enhancement adds a new field `.spec.podReplacementPolicy` to Deployments, to specify whether new Pods can be created only after old Pods have finished terminating, or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can help to better control the speed of rollouts/scaling, and potentially avoid problems with excessive resource usage.
```

I think it is better not to mention scheduling, as this is techincally done by the scheduler and not affected by the deployment controller.

Author: Thank you @atiratree !

> This enhancement adds a new field `.spec.podReplacementPolicy` to Deployments, to specify whether new Pods can be created only after old Pods have finished terminating, or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can help to better control the speed of rollouts/scaling, and potentially avoid problems with excessive resource usage.

Reviewer 5: ```suggestion
This enhancement adds a new field `.spec.podReplacementPolicy` to Deployments to specify whether new Pods can be created only after old Pods have finished terminating or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can better control the speed of rollouts/scaling and potentially avoid problems with excessive resource usage.
```

Author: Implemented, thanks!
